src/machine/asm/execute_x64.S

```diff
@@ -1251,6 +1254,260 @@ ckb_vm_x64_execute:
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  and TEMP1, RS2r
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOOR_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  or TEMP1, RS2r
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOMIN_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  cmp RS2r, TEMP1
+  cmovle TEMP1, RS2r
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOMAX_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  cmp RS2r, TEMP1
+  cmovge TEMP1, RS2r
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOMINU_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  cmp RS2r, TEMP1
+  cmovbe TEMP1, RS2r
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOMAXU_W:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 4)
+  movslq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  WRITE_RD(TEMP1)
+  cmp RS2r, TEMP1
+  cmovae TEMP1, RS2r
+  mov RS2rd, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_LR_D:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  CHECK_READ_VERSION1(RS1, 8)
+  movq CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1), TEMP1
+  movq RS1, LOAD_RESERVATION_ADDRESS
+  WRITE_RD(TEMP1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_SC_D:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_WRITE(RS1, TEMP1d, 8)
+  movq LOAD_RESERVATION_ADDRESS, TEMP1
+  cmp RS1, TEMP1
+  jne .sc_d_branch
+  movq RS2r, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY(MACHINE, RS1)
+  movq $UINT64_MAX, LOAD_RESERVATION_ADDRESS
+  WRITE_RD_VALUE($0, TEMP1)
+  NEXT_INST
+.sc_d_branch:
+  WRITE_RD_VALUE($1, TEMP1)
+  NEXT_INST
+.p2align 3
+.CKB_VM_ASM_LABEL_OP_AMOSWAP_D:
+  DECODE_R
+  movq REGISTER_ADDRESS(RS1), RS1
+  movq REGISTER_ADDRESS(RS2r), RS2r
+  CHECK_READ_VERSION1(RS1, 8)
```

Comment:
For the atomic operations, we will need to check for memory write permissions as well

Reply:
Fixed